arch: reorg stacktrace dependencies & add `arch_stack_walk()` for x86 & arm64

ycsin opened the pull request:
> [!NOTE]
> Expected to conflict with #77457

Fixes #76706
Fixes #74713

Includes #76044 & #76042, obsoletes #75865

___

1. If an arch implements `arch_stack_walk()` `ARCH_HAS_STACKWALK` should be selected.
2. If the above is enabled, `ARCH_SUPPORTS_STACKWALK` indicates if the dependencies are met for arch to enable stack walking. This Kconfig replaces `<arch>_EXCEPTION_STACK_TRACE`
2. If the above is enabled, then, `ARCH_STACKWALK` determines if `arch_stack_walk()` should be compiled.
3. `EXCEPTION_STACK_TRACE` should build on top of the `ARCH_STACKWALK`, stack traces will be printed when it is enabled.
4. `EXCEPTION_STACK_TRACE_MAX_FRAMES ` will be removed as it is replaced by `ARCH_STACKWALK_MAX_FRAMES`

![image](https://github.com/user-attachments/assets/bae9f566-c86f-4054-9f3d-71e8671b10e5)

`CONFIG_EXCEPTION_STACK_TRACE_SYMTAB` is also removed in this PR, as it doesn't really make sense to save a few bytes, when a lot of bytes went into the symbol table when `CONFIG_SYMTAB=y`

ycsin committed:
arch: reorg the dependencies around (exception) stack trace

This commit introduces a new ARCH_STACKWALK Kconfig which
determines if the `arch_stack_walk()` is available should the
arch supports it.

Starting from RISCV, this will be able to converge the exception
stack trace implementation & stack walking features. Existing
exception stack trace implementation will be updated later.
Eventually we will end up with the following:

1. If an arch implements `arch_stack_walk()`
   `ARCH_HAS_STACKWALK` should be selected.
2. If the above is enabled, `ARCH_SUPPORTS_STACKWALK` indicates
   if the dependencies are met for arch to enable stack walking.
   This Kconfig replaces `<arch>_EXCEPTION_STACK_TRACE`
2. If the above is enabled, then, `ARCH_STACKWALK` determines
   if `arch_stack_walk()` should be compiled.
3. `EXCEPTION_STACK_TRACE` should build on top of the
   `ARCH_STACKWALK`, stack traces will be printed when it
   is enabled.
4. `ARCH_STACKWALK_MAX_FRAMES` will be removed as it is
   replaced by `ARCH_STACKWALK_MAX_FRAMES`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
arch: x86: implement `arch_stack_walk()`

Currently it supports `esf` based unwinding only.

Then, update the exception stack unwinding to use
`arch_stack_walk()`, and update the Kconfigs & testcase
accordingly.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
arch: arm64: implement `arch_stack_walk()`

Currently it supports `esf` based unwinding only.

Then, update the exception stack unwinding to use
`arch_stack_walk()`, and update the Kconfigs & testcase
accordingly.

Also, `EXCEPTION_STACK_TRACE_MAX_FRAMES` is unused and
made redundant after this change, so remove it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

dcpleung commented on arch/x86/core/fatal.c:
```diff
@@ -178,16 +178,57 @@ static void unwind_stack(uintptr_t base_ptr, uint16_t cs)
+
 #ifdef CONFIG_X86_64
-		LOG_ERR("     0x%016lx", frame->ret_addr);
+	LOG_ERR("     %d: ra: 0x%016lx", (*i)++, addr);
```
Could you remove the `ra:` part?

ycsin replied:
Removed

ycsin committed:
subsys/debug: remove `CONFIG_EXCEPTION_STACK_TRACE_SYMTAB`

Having `CONFIG_EXCEPTION_STACK_TRACE_SYMTAB` to select the
`CONFIG_SYMTAB` or to explicitly not print the symbol name
during exception stack unwind seems unnecessary, as the extra
code to print the symbol name is negligible when compared with
the symbol table, so just remove it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

dcpleung commented on arch/x86/core/fatal.c:
```diff
@@ -178,16 +178,57 @@ static void unwind_stack(uintptr_t base_ptr, uint16_t cs)
+	LOG_ERR("     %d: 0x%016lx", (*i)++, addr);
 #else
-		LOG_ERR("     0x%08lx (0x%lx)", frame->ret_addr, frame->args);
+	LOG_ERR("     %d: 0x%08lx", (*i)++, addr);
```
Would it  be possible to preserve the printing of `frame->args`?

ycsin replied:
possible if reorg the `unwind_stack()` to use the `walk_stackframe()` instead of `arch_stack_walk()`.

ycsin replied:
@dcpleung I've added a commit for this (the last one), PTAL

nashif commented on arch/riscv/Kconfig:
```diff
@@ -383,6 +376,14 @@ config GEN_IRQ_VECTOR_TABLE
 config ARCH_HAS_SINGLE_THREAD_SUPPORT
 	default y if !SMP
 
+config ARCH_SUPPORTS_STACKWALK
```
why define this per architecture? shouldn't this be in arch/Kconfig and then select it if arch supports it instead?

ycsin replied:
Because each arch seems to have different dependencies

- riscv can be done with or without fp
- x86_32 requires fp, et al
- x86_64 requires fp, `NO_OPTIMIZATIONS`, et al
- arm64 requires fp

this per-arch Kconfig indicates that the dependencies are met and stackwalk can be enabled

But I realize that `ARCH_SUPPORTS_STACKWALK` is redundant, I could have (re)used the `ARCH_HAS_STACKWALK` for the dependency checking

ycsin replied:
> But I realize that `ARCH_SUPPORTS_STACKWALK` is redundant, I could have (re)used the `ARCH_HAS_STACKWALK` for the dependency checking, patch incoming

done

ycsin committed:
arch: document `arch_stack_walk()`

Add `arch_stack_walk()` under the `arch-interface` group and
add documentation on how to add support for it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
arch: riscv: stacktrace: print additional arg when fatal error

Print `sp`/`fp` in traces.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
Squashed the x86 & arm64 's changes in this [commit](https://github.com/zephyrproject-rtos/zephyr/pull/77067#discussion_r1719251121) into where it happened, added a commit to document the API in [Architecture Interface](https://docs.zephyrproject.org/latest/doxygen/html/group__arch-interface.html)

ycsin committed:
doc: releases: update changes about `arch_stack_walk()`

Update changes about `arch_stack_walk()` in RISCV, ARM64 & X86.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
Updated release note
